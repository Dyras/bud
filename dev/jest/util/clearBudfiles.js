const {noop} = require('lodash')
const rimraf = require('rimraf')

module.exports = async () => {
  return rimraf(
    `${process.cwd()}/examples/*/.budfiles`,
    {
      glob: true,
    },
    noop,
  )
}
