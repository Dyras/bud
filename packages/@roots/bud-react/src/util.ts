/**
 * @module @roots/bud-react
 */

import type {Configuration} from 'webpack'

/**
 * @function addRefresh
 */
function addRefresh(entries, [name, assets]) {
  return {
    ...(entries ?? {}),
    [name]: {
      ...assets,
      import: [
        'react-refresh/runtime',
        ...(assets.import ?? []),
      ],
    },
  }
}

/**
 * @function reducer
 */
function reducer(entry: Configuration['entry']) {
  return entry
    ? Object.entries(entry).reduce(addRefresh, entry)
    : {}
}

/**
 * @exports default
 */
export default reducer
