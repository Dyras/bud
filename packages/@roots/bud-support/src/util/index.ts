export {notify} from './notify'
export {dump} from './dump'
export {format} from './format'
export {processHandler} from './processHandler'

import InterpolateHtmlPlugin from './InterpolateHtmlPlugin'
export {InterpolateHtmlPlugin}
