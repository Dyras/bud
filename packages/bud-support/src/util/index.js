"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.InterpolateHtmlPlugin = exports.formatWebpackMessages = exports.eslintFormatter = exports.processHandler = exports.format = exports.dump = exports.notify = void 0;
var notify_1 = require("./notify");
Object.defineProperty(exports, "notify", { enumerable: true, get: function () { return notify_1.notify; } });
var dump_1 = require("./dump");
Object.defineProperty(exports, "dump", { enumerable: true, get: function () { return dump_1.dump; } });
var format_1 = require("./format");
Object.defineProperty(exports, "format", { enumerable: true, get: function () { return format_1.format; } });
var processHandler_1 = require("./processHandler");
Object.defineProperty(exports, "processHandler", { enumerable: true, get: function () { return processHandler_1.processHandler; } });
const eslintFormatter_1 = __importDefault(require("./eslintFormatter"));
exports.eslintFormatter = eslintFormatter_1.default;
const InterpolateHtmlPlugin_1 = __importDefault(require("./InterpolateHtmlPlugin"));
exports.InterpolateHtmlPlugin = InterpolateHtmlPlugin_1.default;
const formatWebpackMessages_1 = __importDefault(require("./formatWebpackMessages"));
exports.formatWebpackMessages = formatWebpackMessages_1.default;
//# sourceMappingURL=index.js.map