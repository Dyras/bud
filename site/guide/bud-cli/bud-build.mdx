---
title: bud build
sidebar_label: bud build
sidebar_position: 2
---

import Tabs from '@theme/Tabs'
import TabItem from '@theme/TabItem'
import CodeBlock from '@theme/CodeBlock'
import BuildHelp from '!!raw-loader!./../../src/components/cli-output/babel-build--help.md'
import Build from '!!raw-loader!./../../src/components/cli-output/babel-build.md'
import BuildCI from '!!raw-loader!./../../src/components/cli-output/babel-build--ci.md'
import BuildCache from '!!raw-loader!./../../src/components/cli-output/babel-build--cache.md'
import BuildHash from '!!raw-loader!./../../src/components/cli-output/babel-build--hash.md'
import BuildMulti from '!!raw-loader!./../../src/components/cli-output/multi-compiler-build.md'
import BuildTargetTheme from '!!raw-loader!./../../src/components/cli-output/multi-compiler-build--target-theme.md'
import BuildTargetThemePlugin from '!!raw-loader!./../../src/components/cli-output/multi-compiler-build--target-theme--target-plugin.md'

<CodeBlock title="$ yarn bud build --help">
  {BuildHelp}
</CodeBlock>

## bud build

<CodeBlock title="$ yarn bud build">{Build}</CodeBlock>

## bud build --cache

The `--cache` flag is recommended for outstanding performance. Compare the `compiled in` output to the uncached build (above).

<CodeBlock title="$ yarn bud build --cache" highlight={6}>
  {BuildCache}
</CodeBlock>

## bud build --target

In multi-compiler configurations, many compilers are run in a batch.

<CodeBlock title="$ yarn bud build">{BuildMulti}</CodeBlock>

The `--target` flag will only compile a specified compiler.

<CodeBlock title="$ yarn bud build --target theme">
  {BuildTargetTheme}
</CodeBlock>

It can be passed more than once.

<CodeBlock title="$ yarn bud build --target theme --target plugin">
  {BuildTargetThemePlugin}
</CodeBlock>

## bud build --hash

The `--hash` flag will enable filename hashing for the build.

<CodeBlock title="$ yarn bud build --hash">
  {BuildHash}
</CodeBlock>

## bud build --ci

The `--ci` flag uses webpack's default cli output (for interoperability).

<CodeBlock title="$ yarn bud build --ci">{BuildCI}</CodeBlock>
