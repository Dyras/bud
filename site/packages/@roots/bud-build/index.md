---
id: "index"
title: "@roots/bud-build"
slug: "/bud-build"
sidebar_label: "Exports"
sidebar_position: 0.5
custom_edit_url: null
---

## Namespaces

- [items](modules/items.md)
- [loaders](modules/loaders.md)
- [rules](modules/rules.md)

## Classes

- [Build](classes/build.md)
- [Item](classes/item.md)
- [Loader](classes/loader.md)
- [Rule](classes/rule.md)

## Functions

### config

â–¸ **config**(`app`): `void`

#### Parameters

| Name | Type |
| :------ | :------ |
| `app` | `Framework` |

#### Returns

`void`

#### Defined in

bud-build/types/Build/config.d.ts:2
