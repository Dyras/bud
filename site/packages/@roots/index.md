---
id: 'packages'
title: '@roots/bud packages'
slug: '/'
sidebar_label: 'Packages'
sidebar_position: 0.5
custom_edit_url: null
---
