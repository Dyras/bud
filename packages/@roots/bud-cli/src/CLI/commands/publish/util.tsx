import {chalk, globby} from '@roots/bud-support'
import {dirname} from 'path'

/**
 * Publishable matches (glob pattern)
 */
export const publishables = [
  ...globby
    .sync([
      '**/node_modules/bud-*/publish/*',
      '**/node_modules/*/bud-*/publish/*',
    ])
    .filter(
      (value, index, self) => self.indexOf(value) === index,
    ),
]

export const formatted = publishables.reduce((a, file) => {
  const ext = dirname(dirname(file)).replace('node_modules/', '')

  const template = file
    .replace(`${ext}/publish/`, '')
    .replace('node_modules/', '')

  return `${a} \n${chalk.blue(ext)} ${chalk.green(template)}`
}, ``)
