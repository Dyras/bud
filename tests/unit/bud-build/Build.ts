import {Item, Loader, Rule} from '@roots/bud-build'

import {Framework, setupBud, teardownBud} from '../../util'

const items = [
  'css',
  'html',
  'style',
  'md',
  'minicss',
  'raw',
  'file',
  'asset',
  'resolve-url',
]

const rules = [
  'css',
  'js',
  'image',
  'font',
  'md',
  'svg',
  'html',
  'xml',
  'json5',
  'yml',
  'yaml',
  'toml',
]

describe('bud.build', function () {
  let bud: Framework

  beforeAll(() => {
    bud = setupBud()
  })

  afterAll(() => {
    bud = teardownBud(bud)
  })

  it('has rebuild method', () => {
    expect(bud.build.rebuild).toBeInstanceOf(Function)
  })

  it('has valid rules', () => {
    Object.entries(bud.build.rules).forEach(([name, rule]) => {
      rules.includes(name)
      expect(rule).toBeInstanceOf(Rule)
    })
  })

  it('has valid items', () => {
    Object.entries(bud.build.items).forEach(([name, item]) => {
      items.includes(name)
      expect(item).toBeInstanceOf(Item)
    })
  })

  it('has valid loaders', () => {
    Object.entries(bud.build.loaders).forEach(
      ([name, loader]) => {
        rules.includes(name)
        expect(loader).toBeInstanceOf(Loader)
      },
    )
  })
})
