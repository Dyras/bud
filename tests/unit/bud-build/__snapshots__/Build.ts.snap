// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bud.build bud.build.config matches snapshot 1`] = `
Object {
  "bail": true,
  "cache": Object {
    "type": "memory",
  },
  "context": Any<String>,
  "devtool": false,
  "entry": undefined,
  "experiments": Object {
    "lazyCompilation": false,
  },
  "externals": undefined,
  "infrastructureLogging": Object {
    "appendOnly": true,
    "level": "none",
  },
  "mode": "production",
  "module": Object {
    "rules": Array [
      Object {
        "parser": Object {
          "requireEnsure": false,
        },
        "test": Object {},
      },
      Object {
        "oneOf": Array [
          Object {
            "exclude": Object {},
            "generator": Object {
              "filename": "assets/[hash][ext][query]",
            },
            "test": Object {},
            "type": "asset/resource",
          },
          Object {
            "exclude": Object {},
            "test": Object {},
            "use": Array [
              Object {
                "loader": StringContaining "resolve-url-loader/index.js",
                "options": Object {
                  "root": StringContaining "src",
                  "sourceMap": false,
                },
              },
            ],
          },
          Object {
            "exclude": Object {},
            "generator": Object {
              "filename": "assets/[hash][ext][query]",
            },
            "test": Object {},
            "type": "asset/resource",
          },
          Object {
            "test": Object {},
            "use": Array [
              Object {
                "loader": StringContaining "html-loader/dist/cjs.js",
              },
            ],
          },
          Object {
            "test": Object {},
            "use": Array [
              Object {
                "loader": StringContaining "csv-loader/index.js",
              },
            ],
          },
          Object {
            "test": Object {},
            "use": Array [
              Object {
                "loader": StringContaining "xml-loader/index.js",
              },
            ],
          },
          Object {
            "parser": Object {
              "parse": Any<Function>,
            },
            "test": Object {},
            "type": "json",
          },
          Object {
            "parser": Object {
              "parse": Any<Function>,
            },
            "test": Object {},
            "type": "json",
          },
          Object {
            "parser": Object {
              "parse": Any<Function>,
            },
            "test": Object {},
            "type": "json",
          },
          Object {
            "exclude": Object {},
            "test": Object {},
            "use": Array [
              Object {
                "loader": StringContaining "mini-css-extract-plugin/dist/loader.js",
                "options": Object {},
              },
              Object {
                "loader": StringContaining "css-loader/dist/cjs.js",
                "options": Object {
                  "importLoaders": 1,
                  "sourceMap": false,
                },
              },
            ],
          },
          Object {
            "exclude": Object {},
            "test": Object {},
            "use": Array [],
          },
        ],
      },
    ],
  },
  "name": "bud",
  "node": false,
  "optimization": Object {
    "emitOnErrors": false,
    "minimize": true,
    "minimizer": Array [
      "...",
      Object {
        "options": Object {
          "exclude": undefined,
          "include": undefined,
          "minify": Any<Function>,
          "minimizerOptions": Object {
            "preset": Array [
              "default",
              Object {
                "discardComments": Object {
                  "removeAll": true,
                },
              },
            ],
          },
          "parallel": true,
          "test": Object {},
          "warningsFilter": Any<Function>,
        },
      },
      Object {
        "options": Object {
          "exclude": undefined,
          "include": undefined,
          "minify": Any<Function>,
          "minimizerOptions": Object {
            "preset": Array [
              "default",
              Object {
                "discardComments": Object {
                  "removeAll": true,
                },
              },
            ],
          },
          "parallel": true,
          "test": Object {},
          "warningsFilter": Any<Function>,
        },
      },
    ],
    "moduleIds": "deterministic",
    "runtimeChunk": undefined,
    "splitChunks": Object {
      "cacheGroups": Object {
        "vendor": Object {
          "chunks": "all",
          "filename": "vendor/[name].js",
          "priority": -10,
          "reuseExistingChunk": true,
          "test": Object {},
        },
      },
    },
  },
  "output": Object {
    "filename": "[name].js",
    "path": StringContaining "dist",
    "pathinfo": undefined,
    "publicPath": "",
  },
  "parallelism": undefined,
  "performance": Object {},
  "plugins": Array [
    Object {
      "apply": Any<Function>,
      "cleanAfterEveryBuildPatterns": Array [],
      "cleanOnceBeforeBuildPatterns": Array [
        "**/*",
        "!dll",
      ],
      "cleanStaleWebpackAssets": true,
      "currentAssets": Array [],
      "dangerouslyAllowCleanPatternsOutsideProject": false,
      "dry": false,
      "handleDone": Any<Function>,
      "handleInitial": Any<Function>,
      "initialClean": false,
      "outputPath": "",
      "protectWebpackAssets": true,
      "removeFiles": Any<Function>,
      "verbose": false,
    },
    Object {
      "depth": 8,
      "includeFalseValues": false,
      "keepCircularReferences": true,
      "name": "bud.webpack.config.js",
      "outputPath": StringContaining ".budfiles",
      "showFunctionNames": true,
    },
    Object {
      "DEBUG": false,
      "ignorePatterns": Array [
        Object {},
      ],
      "options": Object {},
    },
    Object {
      "options": Object {
        "basePath": "",
        "fileName": "manifest.json",
        "filter": null,
        "generate": undefined,
        "map": null,
        "publicPath": "",
        "removeKeyHash": Object {},
        "seed": undefined,
        "serialize": Any<Function>,
        "sort": null,
        "transformExtensions": Object {},
        "useEntryKeys": false,
        "useLegacyEmit": false,
        "writeToFileEmit": true,
      },
    },
    Object {
      "_sortedModulesCache": Any<WeakMap>,
      "options": Object {
        "chunkFilename": "[name].[id].css",
        "experimentalUseImportModule": false,
        "filename": "[name].css",
        "ignoreOrder": false,
      },
      "runtimeOptions": Object {
        "attributes": undefined,
        "insert": undefined,
        "linkType": "text/css",
      },
    },
  ],
  "profile": false,
  "recordsPath": StringContaining ".budfiles/bud-modules.json",
  "resolve": Object {
    "alias": Object {},
    "extensions": Array [
      ".wasm",
      ".mjs",
      ".js",
      ".jsx",
      ".css",
      ".json",
      ".json5",
      ".toml",
      ".xml",
      ".csv",
      ".tsv",
      ".yml",
      ".yaml",
      ".xml",
    ],
    "modules": Array [
      "src",
      "node_modules",
    ],
  },
  "stats": Object {},
  "target": undefined,
  "watch": false,
  "watchOptions": undefined,
}
`;

exports[`bud.build has expected rules 1`] = `
Object {
  "css": Rule {
    "exclude": [Function],
    "generator": undefined,
    "parser": undefined,
    "test": [Function],
    "type": undefined,
    "use": [Function],
  },
  "csv": Rule {
    "exclude": undefined,
    "generator": undefined,
    "parser": undefined,
    "test": [Function],
    "type": undefined,
    "use": [Function],
  },
  "font": Rule {
    "exclude": [Function],
    "generator": undefined,
    "parser": undefined,
    "test": [Function],
    "type": undefined,
    "use": [Function],
  },
  "html": Rule {
    "exclude": undefined,
    "generator": undefined,
    "parser": undefined,
    "test": [Function],
    "type": undefined,
    "use": [Function],
  },
  "image": Rule {
    "exclude": [Function],
    "generator": [Function],
    "parser": undefined,
    "test": [Function],
    "type": [Function],
    "use": undefined,
  },
  "js": Rule {
    "exclude": [Function],
    "generator": undefined,
    "parser": undefined,
    "test": [Function],
    "type": undefined,
    "use": [Function],
  },
  "json5": Rule {
    "exclude": undefined,
    "generator": undefined,
    "parser": [Function],
    "test": [Function],
    "type": [Function],
    "use": undefined,
  },
  "svg": Rule {
    "exclude": [Function],
    "generator": [Function],
    "parser": undefined,
    "test": [Function],
    "type": [Function],
    "use": undefined,
  },
  "toml": Rule {
    "exclude": undefined,
    "generator": undefined,
    "parser": [Function],
    "test": [Function],
    "type": [Function],
    "use": undefined,
  },
  "xml": Rule {
    "exclude": undefined,
    "generator": undefined,
    "parser": undefined,
    "test": [Function],
    "type": undefined,
    "use": [Function],
  },
  "yml": Rule {
    "exclude": undefined,
    "generator": undefined,
    "parser": [Function],
    "test": [Function],
    "type": [Function],
    "use": undefined,
  },
}
`;
